File: answers/leetcode/Single Number III/Single Number III.java
class Solution {
    public int[] singleNumber(int[] nums) {
        int xor = 0;
        for (int num : nums) {
            xor ^= num;
        }
        int bit = xor & -xor;
        int[] result = new int[2];
        for (int num : nums) {
            result[(num & bit) > 0 ? 1 : 0] ^= num;
        }
        return result;
    }
}

File: answers/leetcode/Single Number III/Single Number III.cpp
//@type Bit Manipulation
//@result 30 / 30 test cases passed. Status: Accepted Runtime: 20 ms Submitted: 0 minutes ago You are here! Your runtime beats 23.50% of cpp submissions.

class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        int xors(0);
		for (size_t i = 0; i < nums.size(); ++ i) {
			xors ^= nums[i];
		}
		int pos(0);
		for (int i = 0; i < 32; ++ i) {
			if ((xors >> i) & 1) {
				pos = i;
				break;
			}
		}
		int ones(0);
		int zeros(0);
		for (size_t i = 0; i < nums.size(); ++ i) {
			if ((nums[i] >> pos) & 1) {
				ones ^= nums[i];
			}
			else {
				zeros ^= nums[i];
			}
		}
		vector<int> ans;
		ans.push_back(ones);
		ans.push_back(zeros);
		return ans;
    }
};